### PR TITLE
Added createCfp RPC

### PR DESCRIPTION
**Pull request by Jouzo**
#### What kind of PR is this?:

/kind feature

#### What this PR does / why we need it:

Adds createCfp RPC

#### Which issue(s) does this PR fixes?:

Fixes part of #458

### CODE REVIEW

**Comment by Jouzo** [comment]
Getting npm run test:ci error :
`DeFiDContainer docker not ready within given timeout of 25000ms.
  Unable to find rpc port, the container might have crashed
 `

Could it be linked to fetching other docker image with
`const container = new MasterNodeRegTestContainer(undefined, 'defi/defichain:HEAD-a61f615')`

File: packages/jellyfish-api-core/__tests__/category/governance/createCfp.test.ts
Comment (fuxingloh) [inline packages/jellyfish-api-core/__tests__/category/governance/createCfp.test.ts:36]:
```ts
await container.waitForWalletCoinbaseMaturity() // Wait for wallet balance to increase in order to pay 10 DFI for fund request creation
```

If you need another setup to test `insufficient fund` I think you should have another describe block with another condition: `describe('Governance with insufficient fund', () => {` and setup a new setup of container.

Comment (Jouzo) [reply packages/jellyfish-api-core/__tests__/category/governance/createCfp.test.ts:36]: Ok I can do that, it will be clearer

File: packages/jellyfish-api-core/src/category/governance.ts
Comment (jingyi2811) [inline packages/jellyfish-api-core/src/category/governance.ts:51]: We usually name this UTXO. You can find this interface in other JS files too 

```
export interface UTXO {
  txid: string
  vout: number
}
```

File: packages/jellyfish-api-core/src/category/governance.ts
Comment (jingyi2811) [inline packages/jellyfish-api-core/src/category/governance.ts:33]: Missing `inputs` param

File: packages/jellyfish-api-core/__tests__/category/governance/createCfp.test.ts
Comment (jingyi2811) [inline packages/jellyfish-api-core/__tests__/category/governance/createCfp.test.ts:7]: `without`

File: packages/jellyfish-api-core/__tests__/category/governance/createCfp.test.ts
Comment (jingyi2811) [inline packages/jellyfish-api-core/__tests__/category/governance/createCfp.test.ts:42]: Usually, we add `await container.generate(1)` after executing a RPC to confirm the transaction
You should remove this line

File: packages/jellyfish-api-core/__tests__/category/governance/createCfp.test.ts
Comment (jingyi2811) [inline packages/jellyfish-api-core/__tests__/category/governance/createCfp.test.ts:74]: This should be correct. It confirms the execution of `createCfp` rpc. I think you should add `await container.generate(1)` after` await client.governance.createCfp(data)` for the rest of your code

File: packages/jellyfish-api-core/__tests__/category/governance/createCfp.test.ts
Comment (Jouzo) [reply packages/jellyfish-api-core/__tests__/category/governance/createCfp.test.ts:42]: Without this line, the container stays in 'Initial Block Download'.
Is there a better way to handle this ?

File: packages/jellyfish-api-core/__tests__/category/governance/createCfp.test.ts
Comment (Jouzo) [reply packages/jellyfish-api-core/__tests__/category/governance/createCfp.test.ts:74]: Ok. What about the throwing promises ?

File: packages/jellyfish-api-core/__tests__/category/governance/createCfp.test.ts
Comment (jingyi2811) [reply packages/jellyfish-api-core/__tests__/category/governance/createCfp.test.ts:42]: Try adding it to line 31, which is just after `const promise = client.governance.createCfp(data)`

File: packages/jellyfish-api-core/__tests__/category/governance/createCfp.test.ts
Comment (jingyi2811) [reply packages/jellyfish-api-core/__tests__/category/governance/createCfp.test.ts:74]: I think you should do it too.

Comment (Jouzo) [reply packages/jellyfish-api-core/__tests__/category/governance/createCfp.test.ts:74]: In that case, they don't throw anymore.
From what I can see in the other tests, they don't `await container.generate(1)` to check throwing promises as can be seen in :
`accountToAccount.test.ts` or 
`removePoolLiquidity.test.ts`.
Is there another way to do it ?

File: packages/jellyfish-api-core/__tests__/category/governance/createCfp.test.ts
Comment (Jouzo) [reply packages/jellyfish-api-core/__tests__/category/governance/createCfp.test.ts:42]: In that case, the promises don't throw

File: packages/jellyfish-api-core/__tests__/category/governance/createCfp.test.ts
Comment (jingyi2811) [reply packages/jellyfish-api-core/__tests__/category/governance/createCfp.test.ts:74]: Hmm, you are right. should not apply to promise

File: packages/jellyfish-api-core/__tests__/category/governance/createCfp.test.ts
Comment (jingyi2811) [reply packages/jellyfish-api-core/__tests__/category/governance/createCfp.test.ts:42]: Hmm, you are right. should not apply to promise

Comment (jingyi2811) [reply packages/jellyfish-api-core/__tests__/category/governance/createCfp.test.ts:42]: Without this line, the container stays in 'Initial Block Download'.
Is there a better way to handle this ?

@fuxingloh

Comment (Jouzo) [reply packages/jellyfish-api-core/__tests__/category/governance/createCfp.test.ts:42]: I use this `await container.generate(1)` to generate a block in order get out of Initial Block Download.
What would be the correct way to do it ? @fuxingloh

Comment (fuxingloh) [reply packages/jellyfish-api-core/__tests__/category/governance/createCfp.test.ts:42]: put container.generate in beforeAll

File: packages/jellyfish-api-core/__tests__/category/governance/createCfp.test.ts
Comment (fuxingloh) [inline packages/jellyfish-api-core/__tests__/category/governance/createCfp.test.ts:16]: don't declare global/shared variable when you don't have to, it makes test code harder to read as you need to carry that `let` in your mind creating greater cognitive load when others read your code.

**Comment by jingyi2811** [comment]
I see code like `poolSwapMaxPrice.test.ts ` which is from issue #469 in this PR.
As #469 was merged into main branch just now, can you merge the code from `main` into this branch?

File: packages/jellyfish-api-core/__tests__/category/governance/createCfp.test.ts
Comment (jingyi2811) [inline packages/jellyfish-api-core/__tests__/category/governance/createCfp.test.ts:94]: ` should createCfp with with utxos`

File: packages/jellyfish-api-core/__tests__/category/governance/createCfp.test.ts
Comment (jingyi2811) [inline packages/jellyfish-api-core/__tests__/category/governance/createCfp.test.ts:60]: Rearrange in the order

1) Governance
2) Governance while still in Initial Block Download
3) Governance with insufficient fund

We usually go with positive scenarios first, then only negative scenarios

File: packages/jellyfish-api-core/__tests__/category/governance/createCfp.test.ts
Comment (jingyi2811) [inline packages/jellyfish-api-core/__tests__/category/governance/createCfp.test.ts:136]: Rearrange in this order

1. should createCfp
2. should not createCfp with cycles < 1
3. should not createCfp with cycle > 3
4. should not createCfp with unknown address type
5. should createCfp with with utxos
6. should not createCfp with arbitrary utxos (Add this too)

**Comment by jingyi2811** [comment]
Your PR is polluted. 
it shouldn't have the file like .github/CODEOWNERS 

See `https://github.com/DeFiCh/jellyfish/pull/466/files`

**Comment by Jouzo** [comment]
As draft while waiting for #450 to be merged into main

File: packages/jellyfish-api-core/__tests__/category/governance/createCfp.test.ts
Comment (canonbrother) [inline packages/jellyfish-api-core/__tests__/category/governance/createCfp.test.ts:91]:
```suggestion
    const utxo = await container.fundAddress(payoutAddress, 10)
    const proposalTx = await client.governance.createCfp(data, [utxo])
    await container.generate(1)

    expect(typeof proposalTx).toStrictEqual('string')
    
    const rawtx = await container.call('getrawtransaction', [proposalTx, true])
    expect(rawtx.vin[0].txid).toStrictEqual(utxo.txid)
    expect(rawtx.vin[0].vout).toStrictEqual(utxo.vout)
```

**Review by canonbrother — APPROVED** [review]

File: packages/jellyfish-api-core/src/category/governance.ts
Comment (surangap) [inline packages/jellyfish-api-core/src/category/governance.ts:30]: Mandatory field. remove optional
```suggestion
   * @param {string} data.title Title of community fund request
```

File: packages/jellyfish-api-core/src/category/governance.ts
Comment (surangap) [inline packages/jellyfish-api-core/src/category/governance.ts:32]: same as https://github.com/DeFiCh/jellyfish/pull/466/files#r672185572

File: packages/jellyfish-api-core/src/category/governance.ts
Comment (surangap) [inline packages/jellyfish-api-core/src/category/governance.ts:27]: can add a full stop here. same as the docs
```suggestion
   * Creates a Cummunity Fund Request.
```

File: website/docs/jellyfish/api/governance.md
Comment (jingyi2811) [inline website/docs/jellyfish/api/governance.md:22]: Should be `createCfp (data: CFPData, inputs: UTXO[] = []): Promise<string> {` like what you wrote in the `governance.ts`

File: packages/jellyfish-api-core/src/category/governance.ts
Comment (jingyi2811) [inline packages/jellyfish-api-core/src/category/governance.ts:36]: This should be number

**Review by jingyi2811 — DISMISSED** [review]

**Comment by fuxingloh** [comment]
You can target main branch now

**Comment by Jouzo** [comment]
Had to cleanup after rebase to main

File: packages/testcontainers/src/chains/reg_test_container/masternode.ts
Comment (jingyi2811) [inline packages/testcontainers/src/chains/reg_test_container/masternode.ts:18]: You shouldn't touch this file.

File: packages/jellyfish-api-core/__tests__/category/governance/createCfp.test.ts
Comment (jingyi2811) [inline packages/jellyfish-api-core/__tests__/category/governance/createCfp.test.ts:8]: The existing code definitely allow us to pass in param. This is one of the way

```
 class MyContainer extends MasterNodeRegTestContainer {
    public getCmd (opts: StartOptions): string[] {
      return [
        ...super.getCmd(opts),
        '-fortcanningheight=7'
      ]
    }
  }
  const container = new MyContainer(undefined, 'defi/defichain:HEAD-a61f615', undefined, )

```

Comment (Jouzo) [reply packages/jellyfish-api-core/__tests__/category/governance/createCfp.test.ts]: Indeed, I did a Governance container as it will be needed for following PRs

File: packages/jellyfish-api-core/__tests__/category/governance/createCfp.test.ts
Comment (jingyi2811) [inline packages/jellyfish-api-core/__tests__/category/governance/createCfp.test.ts:161]: For consistency, follow this order -> title, cycles, amount, payoutAddress

**Comment by jingyi2811** [comment]
1 more change request before approving this PR

**Review by jingyi2811 — DISMISSED** [review]

File: website/docs/jellyfish/api/governance.md
Comment (fuxingloh) [inline website/docs/jellyfish/api/governance.md:18]:
```suggestion
Creates a Community Fund Request.
```

File: packages/jellyfish-api-core/src/category/governance.ts
Comment (fuxingloh) [inline packages/jellyfish-api-core/src/category/governance.ts:27]:
```suggestion
   * Creates a Community Fund Request.
```

**Review by fuxingloh — CHANGES_REQUESTED** [review]

**Review by fuxingloh — APPROVED** [review]

**Review by jingyi2811 — APPROVED** [review]

**Pull request merged**
